static/js/narrow.js

```diff
@@ -376,15 +377,18 @@ exports.maybe_add_local_messages = function (opts) {
     }
 
     // Now we know that there are no unread messages, because
-    //   unread_info.flavor === 'not_found'
+    // unread_info.flavor === 'not_found' or this narrow won't
+    // show first unread message.
 
     if (!id_info.target_id) {
         // Without unread messages or a target ID, we're narrowing to
-        // the very latest message matching the narrow.
+        // the very latest message or first unread if matching the narrow allows.
 
-        // TODO: A possible optimization in this code path is to set
-        // `id_info.final_select_id` to be `max_int` here, i.e. saving the
-        // server the first_unread query when we need the server.
+        // Change `id_info.final_select_id` from undefined to `max_int` here,
+        // if matching narrow don't allow to show first unread message.
+        if (!narrow_state.filter().allow_use_first_unread_when_narrowing()) {
+            id_info.final_select_id = 10000000000000000;
+        }
```

Comment:
This change to `maybe_add_local_messages` in several places doesn't feel super simple.  

In any case, I think we should significantly expand the unit tests here.

Reply:
Should i move changes to some other place or just exhaustively test it?